arduino/discovery/discoverymanager/discoverymanager.go

```diff
@@ -60,234 +79,174 @@ func (dm *DiscoveryManager) IDs() []string {
-}
-
-// List returns a list of available ports detected from all discoveries
-// and a list of errors for those discoveries that returned one.
-func (dm *DiscoveryManager) List() ([]*discovery.Port, []error) {
-	var wg sync.WaitGroup
-	// Use this struct to avoid the need of two separate
-	// channels for ports and errors.
-	type listMsg struct {
-		Err  error
-		Port *discovery.Port
+	if err := d.Run(); err != nil {
+		return fmt.Errorf(tr("discovery %[1]s process not started: %[2]w"), d.GetID(), err)
 	}
-	msgChan := make(chan listMsg)
-	dm.discoveriesMutex.Lock()
-	discoveries := []*discovery.PluggableDiscovery{}
-	for _, d := range dm.discoveries {
-		discoveries = append(discoveries, d)
-	}
-	dm.discoveriesMutex.Unlock()
-	for _, d := range discoveries {
-		wg.Add(1)
-		go func(d *discovery.PluggableDiscovery) {
-			defer wg.Done()
-			if d.State() != discovery.Running {
-				// Discovery is not running, it won't return anything
-				return
-			}
-			ports, err := d.List()
-			if err != nil {
-				msgChan <- listMsg{Err: fmt.Errorf(tr("listing ports from discovery %[1]s: %[2]w"), d.GetID(), err)}
-			}
-			for _, p := range ports {
-				msgChan <- listMsg{Port: p}
-			}
-		}(d)
+	eventCh, err := d.StartSync(5)
+	if err != nil {
+		return fmt.Errorf(tr("start syncing discovery %[1]s: %[2]w"), d.GetID(), err)
 	}
 
+	// XXX do better cleanup if the discovery fails to start
+
 	go func() {
-		// Close the channel only after all goroutines are done
-		wg.Wait()
-		close(msgChan)
+		// Transfer all incoming events from this discovery to the feed channel
+		for ev := range eventCh {
+			dm.feed <- ev
+		}
 	}()
+	return nil
+}
+
+func (dm *DiscoveryManager) feedEvent(ev *discovery.Event) {
+	dm.watchersMutex.Lock()
+	defer dm.watchersMutex.Unlock()
 
-	ports := []*discovery.Port{}
-	errs := []error{}
-	for msg := range msgChan {
-		if msg.Err != nil {
-			errs = append(errs, msg.Err)
-		} else {
-			ports = append(ports, msg.Port)
+	if ev.Type == "stop" {
+		// Remove all the cached events for the terminating discovery
+		delete(dm.watchersCache, ev.DiscoveryID)
+		return
+	}
+
+	// Send the event to all watchers
+	for watcher := range dm.watchers {
+		select {
+		case watcher.feed <- ev:
+			// OK
+		case <-time.After(time.Millisecond * 500):
+			// If the watcher is not able to process event fast enough
+			// remove the watcher from the list of watchers
+			logrus.Info("Watcher is not able to process events fast enough, removing it from the list of watchers")
+			delete(dm.watchers, watcher)
 		}
 	}
-	return ports, errs
+
+	// Cache the event for the discovery
+	cache := dm.watchersCache[ev.DiscoveryID]
+	if cache == nil {
+		cache = map[string]*discovery.Event{}
+		dm.watchersCache[ev.DiscoveryID] = cache
+	}
+	eventID := ev.Port.Address + "|" + ev.Port.Protocol
+	switch ev.Type {
+	case "add":
+		cache[eventID] = ev
+	case "remove":
+		delete(cache, eventID)
+	default:
+		logrus.Errorf("Unhandled event from discovery: %s", ev.Type)
+	}
```

Comment:
During the review you told me that maybe this can be removed, have you had the chance to verify this?

Reply:
I removed it in the latest push,